core/test/mpi/base/bindings.cpp

```diff
@@ -0,0 +1,782 @@
+            gko::Array<ValueType>{ref, gko::Array<ValueType>(ref, 4, data)};
+        for (auto rank = 0; rank < num_ranks; ++rank) {
+            if (rank != my_rank) {
+                gko::mpi::send<ValueType>(send_array.get_const_data(), 4, rank,
+                                          40 + rank, comm);
+            }
+        }
+    } else {
+        recv_array = gko::Array<ValueType>{ref, 4};
+        gko::mpi::recv<ValueType>(recv_array.get_data(), 4, 0, 40 + my_rank,
+                                  comm);
+    }
+    if (my_rank != 0) {
+        ASSERT_EQ(recv_array.get_data()[0], 1);
+        ASSERT_EQ(recv_array.get_data()[1], 2);
+        ASSERT_EQ(recv_array.get_data()[2], 3);
+        ASSERT_EQ(recv_array.get_data()[3], 4);
+    }
+}
+
+
+TEST_F(MpiBindings, CanNonBlockingSendAndNonBlockingRecvValues)
+{
+    using ValueType = int;
+    auto comm = gko::mpi::communicator::create_world();
+    auto my_rank = comm->rank();
+    auto num_ranks = comm->size();
+    auto send_array = gko::Array<ValueType>{ref};
+    auto recv_array = gko::Array<ValueType>{ref};
+    ValueType* data;
+    auto req = gko::mpi::request::create(num_ranks);
+    if (my_rank == 0) {
+        data = new ValueType[4]{1, 2, 3, 4};
+        send_array =
+            gko::Array<ValueType>{ref, gko::Array<ValueType>(ref, 4, data)};
+        for (auto rank = 0; rank < num_ranks; ++rank) {
+            if (rank != my_rank) {
+                gko::mpi::send<ValueType>(send_array.get_data(), 4, rank,
+                                          40 + rank, req, comm);
+            }
+        }
+    } else {
+        recv_array = gko::Array<ValueType>{ref, 4};
+        gko::mpi::recv<ValueType>(recv_array.get_data(), 4, 0, 40 + my_rank,
+                                  req, comm);
+    }
+    gko::mpi::wait(req);
+    if (my_rank != 0) {
+        ASSERT_EQ(recv_array.get_data()[0], 1);
+        ASSERT_EQ(recv_array.get_data()[1], 2);
+        ASSERT_EQ(recv_array.get_data()[2], 3);
+        ASSERT_EQ(recv_array.get_data()[3], 4);
+    }
+}
+
+
+TEST_F(MpiBindings, CanPutValuesWithLockAll)
+{
+    using ValueType = int;
+    using window = gko::mpi::window<ValueType>;
+    auto comm = gko::mpi::communicator::create_world();
+    auto my_rank = comm->rank();
+    auto num_ranks = comm->size();
+    int* data;
+    if (my_rank == 0) {
+        data = new ValueType[4]{1, 2, 3, 4};
+    } else {
+        data = new ValueType[4]{0, 0, 0, 0};
+    }
+    auto win = window(data, 4 * sizeof(ValueType), comm);
+    win.lock_all();
+    if (my_rank == 0) {
+        for (auto rank = 0; rank < num_ranks; ++rank) {
+            if (rank != my_rank) {
+                gko::mpi::put<ValueType>(data, 4, rank, 0, 4, win);
+                win.flush(rank);
+            }
+        }
+    }
+    win.unlock_all();
+    gko::mpi::synchronize(comm);
+    ASSERT_EQ(data[0], 1);
+    ASSERT_EQ(data[1], 2);
+    ASSERT_EQ(data[2], 3);
+    ASSERT_EQ(data[3], 4);
+    delete data;
```

Comment:
I think you should use a `std::vector` for allocating these buffers.
```suggestion
    delete [] data;
```